src/pset/mod.rs

```diff
@@ -248,13 +253,296 @@ impl PartiallySignedTransaction {
+    }
+
+    /// Blind the pset as the non-last blinder role. The last blinder of pset
+    /// should call the `blind_last` function which balances the blinding factors
+    /// `inp_secrets` and must be consistent by [`Output`] `blinder_index` field
+    /// For each output that is to be blinded, the following must be true
+    /// 1. The blinder_index must be set in pset output field
+    /// 2. the corresponding inp_secrets\[out.blinder_index\] must be present
+    /// # Parameters
+    ///
+    /// * `inp_secrets`: [`TxOutSecrets`] corresponding to owned inputs. Use [`None`] for non-owned outputs
+    ///
+    pub fn blind_non_last<C: secp256k1_zkp::Signing, R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        secp: &secp256k1_zkp::Secp256k1<C>,
+        inp_txout_sec: &[Option<&TxOutSecrets>],
+    ) -> Result<Vec<(AssetBlindingFactor, ValueBlindingFactor)>, PsetBlindError> {
+        let (inp_secrets, outs_to_blind) = self.blind_checks(inp_txout_sec)?;
+
+        // Blind each output as non-last and save the secrets
+        let spent_utxos = self
+            .inputs
+            .iter()
+            .enumerate()
+            .map(|(i, x)| x.witness_utxo.as_ref().ok_or(PsetBlindError::MissingWitnessUtxo(i)))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let spent_utxo_secrets = spent_utxos
+            .iter()
+            .map(|x| x.asset)
+            .zip(inp_txout_sec.iter().map(|x| *x))
+            .collect::<Vec<(_, _)>>();
+        let mut out_secrets = vec![];
+        let mut ret = vec![]; // return all the random values used
+        for i in outs_to_blind {
+            let mut txout = self.outputs[i].to_txout();
+            let (abf, vbf) = txout
+                .to_non_last_confidential(
+                    rng,
+                    secp,
+                    self.outputs[i]
+                        .blinding_key
+                        .map(|x| x.key)
+                        .ok_or(PsetBlindError::MustHaveExplicitTxOut(i))?,
+                    &spent_utxo_secrets,
+                )
+                .unwrap();
+            let value = self.outputs[i].amount.explicit().ok_or(PsetBlindError::MustHaveExplicitTxOut(i))?;
+            out_secrets.push((value, abf, vbf));
+
+            // mutate the pset
+            {
+                self.outputs[i].value_rangeproof = txout.witness.rangeproof;
+                self.outputs[i].asset_surjection_proof = txout.witness.surjection_proof;
+                self.outputs[i].amount = txout.value;
+                self.outputs[i].asset = txout.asset;
+                self.outputs[i].ecdh_pubkey = txout.nonce.commitment().map(|pk| bitcoin::PublicKey{
+                    key: pk,
+                    compressed: true
+                });
+            }
+            // return blinding factors used
+            ret.push((abf, vbf));
+        }
+
+        let (value, abf, vbf) = *out_secrets.last().unwrap();
```

Comment:
Here we can use `.pop()` rather than `.last()`, and then below we don't need to `len() - 1`.

Then to avoid issues in the case that there are no outputs to be blinded, we can do `unwrap_or(0, ValueBlindingFactor::zero(), ValueBlindingFactor::zero())`.

Reply:
err `unwrap_or((0, ZERO_TWEAK, ZERO_TWEAK))`

Reply:
I have added a early return if outs_to_blind = 0, so can safely unwrap() here.

Reply:
It is more idiomatic to eliminate the `unwrap`s, and if we're early-returning rather than panicking, to do this directly rather than having a separate branch

It is also more idiomatic to just use `pop` rather than using `last` and then using extra arithmetic to ignore the last item.